### DISCUSSION TITLE
[request] Environment for QMK

### DISCUSSION THREAD

**Discussion by AndersonTorres** [nix-community/nix-environments#19] [Ideas]
Copy-pasting https://github.com/NixOS/nixpkgs/issues/129065:

> QMK is already packaged into Nixpkgs (#127863). However, it does not work standalone - it is a tool made to manipulate QMK firmware source tree.

> Therefore, its best use case is inside a well-crafted shell.nix file.

Reference: https://github.com/NixOS/nixpkgs/issues/134530

**Comment by Mic92** [comment]
If it is packaged, can you not just use `nix-shell <nixpkgs> -A qmk` to get a build environment?

**Comment by AndersonTorres** [comment]
> If it is packaged, can you not just use `nix-shell <nixpkgs> -A qmk` to get a build environment?

It is a bit complicated to explain.

QMK, as packaged by Nixpkgs, is mostly useful for download and manipulate the QMK Firmware repo. However, the other functionalities, especially compile and burn the firmware to a keyboard, require extra packages like cross-compilers and firmware loaders.

The "expected workflow" for QMK is something like: 

1. Download the qmk firmware git repo - here you use `qmk` tool
2. Edit the source code to match your needs (alternative keyboard, blinking etc.) - here you use emacs or mousepad :)
3. Compile the firmware - here you can use `qmk` tool if the AVR cross-compiler is found on `$PATH`
4. Burn it to the keyboard - here you can use `qmk` tool if avrdude is found on `$PATH`

You can think on it as something like Arduino programming (in certain sense it is).

**Reply by zarelit** [reply]
Out of ignorance: isn't this what's brought in by the `shell.nix` present in the qmk_firmware repository?

**Reply by AndersonTorres** [reply]
According to some users, this `shell.nix` file from `qmk_firmware` repo is out of date and it isn't maintained, not being a huge concern for QMK project.

**Reply by RealET** [reply]
Which AVR cross-compiler are you using?

**Comment by Mic92** [comment]
Ok. Pull requests are welcome. I will no likely add this as I don't use this.
